[Do not merge] Fix nested array copy from calldata to storage

sifmelcara opened the pull request:
### Description

This PR is similar to #6881 , but fixes nested calldata dynamic array's assignment to storage. (#6881 fixes assignment to memory)

The `Changelog.md` update reflects both this PR and #6881 's change.

I only tested some basic cases so please review carefully.

By the way, the codegen of other dynamic types that need calldata tail pointer access are guarded by `solUnimplementedAssert`, and for array types it should always cause stack height assertion failure (because calldata dynamic array's length is also stored on stack). So the bug fixed by these 2 PRs probably did not lead to vulnerability in previous compiler releases.

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

chriseth commented:
I'd like to request another review by @ekpyron before this can be merged.

ekpyron reviewed (commented):
Phew - these routines are extremely complicated - I wonder whether we should just delegate all the complex multi-dimensional ABIEncoderV2 cases to Yul functions instead...
But again: on a casual look this makes sense, but I'd also recommend to add a test for three-dimensional arrays similar to #6881 - it should work in the same way without too much hassle.

chriseth commented:
I think it would still be better to wait for you to return.

sifmelcara commented:
(just forced-pushed to rebase)

sifmelcara committed:
Add tests that copys nested calldata array to storage

sifmelcara committed:
Fix nested dynamic calldata array assignment to storage

Fix assertion error (stack height mismatch) happend when copy nested
dynamic calldata array to storage.

sifmelcara commented:
Force pushed to address an issue about I incorrectly assumed dynamically encoded array must be dynamically sized. (Which is not true during codegen 3-dimensional array assignment)

But there are still an issue about source pointer advancement, described in #7072

ekpyron commented:
@sifmelcara Now that https://github.com/ethereum/solidity/pull/7178 is merged, we could rebase this and check whether it solves the issue with three-dimensional arrays!

sifmelcara committed:
Add three-dimensional nested array calldata to storage assignment test

sifmelcara commented:
:tada: It now passes after rebase.
Now what remained to be done is to rebase again after 0.5.11 release (because of changelog.md) and squash commits.

ekpyron commented:
The failing test in the t_ubu_asan_constantinople may be unrelated to this PR and rather due to the macros disabling that check in https://github.com/ethereum/solidity/pull/7010 not working...

ekpyron commented:
We decided that we don't want to properly implement this for the old ABIEncoder, but instead want to push ABIEncoderV2 forward, so I'm closing this. I'll open a new issue for improving error messages in these cases.
